src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+            start(js);
+          }
+        }
+        KJ_CASE_ONEOF(started, Started) {
+          // If we are in the started state, stop the port if there are no listeners.
+          if (count == 0 && onmessageValue == kj::none) {
+            state = Pending();
+          }
+        }
+        KJ_CASE_ONEOF(_, Closed) {
+          // Nothing to do. We're already closed so we don't care.
+        }
+      }
+    }
+  });
+}
+
+// Deliver the message to the "message" or "messageerror" event on this port.
+void MessagePort::deliver(
+    jsg::Lock& js, kj::Own<rpc::JsValue::Reader> message, jsg::Ref<MessagePort> port) {
+  auto event = js.tryCatch([&] {
+    jsg::Deserializer deserializer(js, message->getV8Serialized());
+    return MessageEvent::New(
+        js, jsg::JsRef(js, deserializer.readValue(js)), kj::str(), port->addRef(), {});
+  }, [&](jsg::Value exception) {
+    return MessageEvent::NewError(
+        js, jsg::JsRef(js, jsg::JsValue(exception.getHandle(js))), kj::str(), port->addRef(), {});
+  });
+
+  // If any of the message/messageerror event handlers throw,
+  // capture the error and pass it to reportError instead of
+  // propagating up.
+  js.tryCatch([&] { port->dispatchEventImpl(js, kj::mv(event)); }, [&](jsg::Value exception) {
+    auto context = js.v8Context();
+    auto& global =
+        jsg::extractInternalPointer<ServiceWorkerGlobalScope, true>(context, context->Global());
+    global.reportError(js, jsg::JsValue(exception.getHandle(js)));
+  });
+}
+
+// Deliver the message to all the jsrpc remotes we have
+kj::Promise<void> MessagePort::sendToRpc(kj::Own<rpc::JsValue::Reader> message) {
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+  kj::Vector<kj::Promise<void>> promises;
+  for (rpc::JsMessagePort::Client cap: rpcClients) {
+    auto req = cap.callRequest();
+    req.setData(*message);
+    // TODO(message-port): Removing the port when dropped?
+    promises.add(req.send().ignoreResult());
+  }
+  co_await kj::joinPromises(promises.releaseAsArray());
+}
+
+// Deliver the message to this port, buffering if necessary if the port
+// has not been started. Buffered messages will be delivered when the
+// port is started later.
+void MessagePort::deliverMessage(jsg::Lock& js, rpc::JsValue::Reader value) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(pending, Pending) {
+      // We have not yet started the port so buffer to message.
+      pending.add(capnp::clone(value));
+    }
+    KJ_CASE_ONEOF(started, Started) {
+      ioContext.addTask(sendToRpc(capnp::clone(value)));
+      deliver(js, capnp::clone(value), addRef());
+    }
+    KJ_CASE_ONEOF(_, Closed) {
+      // Nothing to do in this case. Drop the message on the floor.
+    }
+  }
+}
+
+// Binds two ports to each other such that messages posted to one
+// are delivered on the other.
+void MessagePort::entangle(MessagePort& port1, MessagePort& port2) {
+  port1.other = port2.addRef();
```

Comment:
This has a bug. If this port is entangled with another port, we should disentangle first. If not, port3 that is entangled with port1 will still be entangled with port1, even though port1 is entangled with port2.

Step 1 of entangle spec step: https://html.spec.whatwg.org/multipage/web-messaging.html#entangle

Reply:
Also disentangle should emit "close" event which in this case it never emits.

Reply:
That shouldn't ever happen. If anything, this should just assert if the port is already entangled, but in every use of this currently there's no way for the ports to already be entangled.